Regression: Type cannot be used to index type 'this'.

<!-- 🚨 STOP 🚨 𝗦𝗧𝗢𝗣 🚨 𝑺𝑻𝑶𝑷 🚨

Half of all issues filed here are duplicates, answered in the FAQ, or not appropriate for the bug tracker. Even if you think you've found a *bug*, please read the FAQ first, especially the Common "Bugs" That Aren't Bugs section!

Please help us by doing the following steps before logging an issue:
  * Search: https://github.com/Microsoft/TypeScript/search?type=Issues
  * Read the FAQ: https://github.com/Microsoft/TypeScript/wiki/FAQ

Please fill in the *entire* template below.
-->

<!-- Please try to reproduce the issue with `typescript@next`. It may have already been fixed. -->
**TypeScript Version:**  3.5.0-dev.20190517

<!-- Search terms you tried before logging this (so others can find this issue more easily) -->
**Search Terms:**
TS2536
Type cannot be used to index type 'this'.

**Code**

```ts
export class c {
  [x: string]: string;
  constructor() {
    this["a"] = "b";
  }
}
```

**Expected behavior:**
No errors

**Actual behavior:**
error TS2536: Type '"a"' cannot be used to index type 'this'.

**Playground Link:** <!-- A link to a TypeScript Playground "Share" link which demonstrates this behavior -->
Works fine in playground but fails in 3.5.0.rc
https://www.typescriptlang.org/play/#src=class%20y%20%7B%0D%0A%20%20%20%20%5Bx%3A%20string%5D%3A%20string%3B%0D%0A%20%20%20%20constructor()%20%7B%0D%0A%20%20%20%20%20%20%20%20this%5B%22b%22%5D%20%3D%20%22a%22%3B%0D%0A%20%20%20%20%7D%0D%0A%7D%0D%0A

**Related Issues:** <!-- Did you find other bugs that looked similar? -->
Could not find any

--- a/src/compiler/checker.ts
+++ b/src/compiler/checker.ts
@@ -3563,7 +3563,7 @@ namespace ts {
                     context.approximateLength += 6;
                     return createKeywordTypeNode(SyntaxKind.ObjectKeyword);
                 }
-                if (type.flags & TypeFlags.TypeParameter && (type as TypeParameter).isThisType) {
+                if (isThisTypeParameter(type)) {
                     if (context.flags & NodeBuilderFlags.InObjectTypeLiteral) {
                         if (!context.encounteredError && !(context.flags & NodeBuilderFlags.AllowThisInObjectLiteral)) {
                             context.encounteredError = true;
@@ -10193,6 +10193,10 @@ namespace ts {
             return maybeTypeOfKind(type, TypeFlags.InstantiableNonPrimitive | TypeFlags.Index);
         }
 
+        function isThisTypeParameter(type: Type): boolean {
+            return !!(type.flags & TypeFlags.TypeParameter && (<TypeParameter>type).isThisType);
+        }
+
         function getSimplifiedType(type: Type, writing: boolean): Type {
             return type.flags & TypeFlags.IndexedAccess ? getSimplifiedIndexedAccessType(<IndexedAccessType>type, writing) :
                 type.flags & TypeFlags.Conditional ? getSimplifiedConditionalType(<ConditionalType>type, writing) :
@@ -16772,7 +16776,7 @@ namespace ts {
             }
 
             function narrowByInKeyword(type: Type, literal: LiteralExpression, assumeTrue: boolean) {
-                if ((type.flags & (TypeFlags.Union | TypeFlags.Object)) || (type.flags & TypeFlags.TypeParameter && (type as TypeParameter).isThisType)) {
+                if (type.flags & (TypeFlags.Union | TypeFlags.Object) || isThisTypeParameter(type)) {
                     const propName = escapeLeadingUnderscores(literal.text);
                     return filterType(type, t => isTypePresencePossible(t, propName, assumeTrue));
                 }
@@ -20092,7 +20096,7 @@ namespace ts {
                         return anyType;
                     }
                     if (right.escapedText && !checkAndReportErrorForExtendingInterface(node)) {
-                        reportNonexistentProperty(right, leftType.flags & TypeFlags.TypeParameter && (leftType as TypeParameter).isThisType ? apparentType : leftType);
+                        reportNonexistentProperty(right, isThisTypeParameter(leftType) ? apparentType : leftType);
                     }
                     return errorType;
                 }
@@ -20496,7 +20500,7 @@ namespace ts {
 
             const effectiveIndexType = isForInVariableForNumericPropertyNames(indexExpression) ? numberType : indexType;
             const accessFlags = isAssignmentTarget(node) ?
-                AccessFlags.Writing | (isGenericObjectType(objectType) ? AccessFlags.NoIndexSignatures : 0) :
+                AccessFlags.Writing | (isGenericObjectType(objectType) && !isThisTypeParameter(objectType) ? AccessFlags.NoIndexSignatures : 0) :
                 AccessFlags.None;
             const indexedAccessType = getIndexedAccessTypeOrUndefined(objectType, effectiveIndexType, node, accessFlags) || errorType;
             return checkIndexedAccessIndexType(indexedAccessType, node);

--- a/tests/baselines/reference/keyofAndIndexedAccess2.errors.txt
+++ b/tests/baselines/reference/keyofAndIndexedAccess2.errors.txt
@@ -220,3 +220,12 @@ tests/cases/conformance/types/keyof/keyofAndIndexedAccess2.ts(108,5): error TS23
         let y: ReadonlyArray<string>[K] = 'abc';
     }
     
+    // Repro from #31439
+    
+    export class c {
+      [x: string]: string;
+      constructor() {
+        this["a"] = "b";
+      }
+    }
+    

--- a/tests/baselines/reference/keyofAndIndexedAccess2.js
+++ b/tests/baselines/reference/keyofAndIndexedAccess2.js
@@ -137,6 +137,15 @@ function fn4<K extends number>() {
     let y: ReadonlyArray<string>[K] = 'abc';
 }
 
+// Repro from #31439
+
+export class c {
+  [x: string]: string;
+  constructor() {
+    this["a"] = "b";
+  }
+}
+
 
 //// [keyofAndIndexedAccess2.js]
 function f1(obj, k0, k1, k2) {
@@ -225,4 +234,10 @@ function fn3(param, cb) {
 function fn4() {
     let x = 'abc';
     let y = 'abc';
+}
+// Repro from #31439
+export class c {
+    constructor() {
+        this["a"] = "b";
+    }
 }

--- a/tests/baselines/reference/keyofAndIndexedAccess2.symbols
+++ b/tests/baselines/reference/keyofAndIndexedAccess2.symbols
@@ -503,3 +503,17 @@ function fn4<K extends number>() {
 >K : Symbol(K, Decl(keyofAndIndexedAccess2.ts, 133, 13))
 }
 
+// Repro from #31439
+
+export class c {
+>c : Symbol(c, Decl(keyofAndIndexedAccess2.ts, 136, 1))
+
+  [x: string]: string;
+>x : Symbol(x, Decl(keyofAndIndexedAccess2.ts, 141, 3))
+
+  constructor() {
+    this["a"] = "b";
+>this : Symbol(c, Decl(keyofAndIndexedAccess2.ts, 136, 1))
+  }
+}
+

--- a/tests/baselines/reference/keyofAndIndexedAccess2.types
+++ b/tests/baselines/reference/keyofAndIndexedAccess2.types
@@ -499,3 +499,21 @@ function fn4<K extends number>() {
 >'abc' : "abc"
 }
 
+// Repro from #31439
+
+export class c {
+>c : c
+
+  [x: string]: string;
+>x : string
+
+  constructor() {
+    this["a"] = "b";
+>this["a"] = "b" : "b"
+>this["a"] : string
+>this : this
+>"a" : "a"
+>"b" : "b"
+  }
+}
+

--- a/tests/cases/conformance/types/keyof/keyofAndIndexedAccess2.ts
+++ b/tests/cases/conformance/types/keyof/keyofAndIndexedAccess2.ts
@@ -138,3 +138,12 @@ function fn4<K extends number>() {
     let x: Array<string>[K] = 'abc';
     let y: ReadonlyArray<string>[K] = 'abc';
 }
+
+// Repro from #31439
+
+export class c {
+  [x: string]: string;
+  constructor() {
+    this["a"] = "b";
+  }
+}

